pkg/controller/pipelinerun/controller.go

```diff
@@ -0,0 +1,241 @@
+		pipelinesLister: pipelinesListerFunc(informer),
+		synced:          runsInformer.Informer().HasSynced,
+		workqueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PipelineRuns"),
+		recorder:        getRecordFunc(k, l),
+		logger:          l,
+	}
+
+	l.Info("Setting up event handlers")
+	// Set up an event handler for when PipelineRun resource is added
+	runsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueuePipelineRun,
+	})
+	return controller
+}
+
+// Run will set up the event handlers for types we are interested in, as well
+// as syncing informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	// Start the informer factories to begin populating the informer caches
+	c.logger.Info("Starting PipelineRun controller")
+
+	// Wait for the caches to be synced before starting workers
+	c.logger.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.synced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	c.logger.Info("Starting workers")
+	// Launch two workers to process Build resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	c.logger.Info("Started workers")
+	<-stopCh
+	c.logger.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	if err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		key, ok := obj.(string)
+		if !ok {
+			c.workqueue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := c.syncHandler(key); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", key, err.Error())
+		}
+		c.workqueue.Forget(obj)
+		c.logger.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj); err != nil {
+		runtime.HandleError(err)
+	}
+
+	return true
+}
+
+// enqueuePipelineRun takes a PipelineRun resource and converts it into a namespace/name
+// string which is then put onto the work queue.
+func (c *Controller) enqueuePipelineRun(obj interface{}) {
+	var key string
+	var err error
```

Comment:
do these both need to be declared here or can we use the `:=` syntax?

Reply:
@bobcatfish yes, as `key` is used outside it's scope otherwise. An alternative would be

```go
if key, err := cache.MetaNamespaceKeyFunc(obj); err != nil {
	runtime.HandleError(err)
	return
} else {
	c.workqueue.AddRateLimited(key)
}
```

Reply:
ah, it's because of the 1 line err handling syntax isn't it! got it, thanks. ill get used to that syntax one of these days.

Reply:
:)